1a1a11a/cleanup CMakeList

1a1a11a opened the pull request.

1a1a11a committed:
cleanup cmakelists

1a1a11a committed:
add dev env install

haochengxia commented on scripts/install_dependency.sh:
```diff
@@ -223,6 +223,13 @@ install_zstd() {
     popd > /dev/null
 }
 
+install_dev() {
```
Do we want to remove dev dep like `clang-tidy` from `setup_ubuntu` and only preserve it in `install_dev`? Now both of them have it.

1a1a11a replied:
will update

1a1a11a commented:
This one depends on #201.

1a1a11a committed:
cleanup continue

1a1a11a committed:
cleanup library installation

1a1a11a committed:
remove admission and prefetch link library

1a1a11a committed:
update eviction camke
remove private

1a1a11a committed:
fix clang-tidy

1a1a11a committed:
separate dev dependency; reformat

haochengxia commented:
Some thoughts:

1. Is it necessary to reduce the usage of global `set`?
    e.g.,
    ```
    set(LIBS ${LIBS} ${GLib_LIBRARY})
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wextra ...")
    
    ->
    
    target_compile_options(${PROJECT_NAME} PRIVATE -Wall -Wextra -Werror ...)
    target_link_libraries(${PROJECT_NAME} PRIVATE ${GLib_LIBRARIES})
    ```

2. Should we use `set(... PARENT_SCOPE)` in subdir's CMakeLists.txt to expose the file to compile in subdir instead of organize the code via manually gathering all source code in the main CMakeLists.txt?
    e.g.,
    ```
    # libCacheSim/dataStructure/CMakeLists.txt
    
    set(DATASTRUCTURE_SOURCES
        pqueue.c
        splay.c
        bloom.c
        minimalIncrementCBF.c
        histogram.c
        splay_tuple.c
       ...
        PARENT_SCOPE
    )
    
    # CMakeLists.txt
    add_library(${PROJECT_NAME} 
        ${DATASTRUCTURE_SOURCES}
    ...)
    
    ```
4. Do we want to export ````libCacheSimConfig.cmake````?

1a1a11a committed:
many fixes

1a1a11a commented:
> Some thoughts:
> 
> 1. Is it necessary to reduce the usage of global `set`?
>    e.g.,
>    ```
>    set(LIBS ${LIBS} ${GLib_LIBRARY})
>    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wextra ...")
>    
>    ->
>    
>    target_compile_options(${PROJECT_NAME} PRIVATE -Wall -Wextra -Werror ...)
>    target_link_libraries(${PROJECT_NAME} PRIVATE ${GLib_LIBRARIES})
>    ```
> 2. Should we use `set(... PARENT_SCOPE)` in subdir's CMakeLists.txt to expose the file to compile in subdir instead of organize the code via manually gathering all source code in the main CMakeLists.txt?
>    e.g.,
>    ```
>    # libCacheSim/dataStructure/CMakeLists.txt
>    
>    set(DATASTRUCTURE_SOURCES
>        pqueue.c
>        splay.c
>        bloom.c
>        minimalIncrementCBF.c
>        histogram.c
>        splay_tuple.c
>       ...
>        PARENT_SCOPE
>    )
>    
>    # CMakeLists.txt
>    add_library(${PROJECT_NAME} 
>        ${DATASTRUCTURE_SOURCES}
>    ...)
>    ```
> 3. Do we want to export `libCacheSimConfig.cmake`?

1 . good point, updated. I used a global variable to allow use change, but I set the property on each lib, so we can change independently. 
2. I agree, updated and also cleaned up more, which causes the changes in this PR to be huge
3. added `libCacheSimTargets.cmake` to cmake

because I added a global include, so I will also update the various header include not to use relative path. 
But I will do it in the next PR to avoid making too many changes in this PR...

1a1a11a committed:
fix bugs and disable mrcProfiler test

1a1a11a commented:
I temporarily removed mrcProfiler from the test because the plugin system needs more work and will be addressed separately.

1a1a11a committed:
fix formatting

1a1a11a commented:
@haochengxia ready to review

1a1a11a committed:
fix thread init issue

1a1a11a committed:
add export-dynamic to cachesim

1a1a11a committed:
fix a double-free bug

haochengxia reviewed (approved):
LGTM!

P.S.: I noticed that in some places, such as `admission`, we used `include_directories` to include header files, so we no longer need relative paths. Should we apply this to the project scope? If yes, I'd like to create one PR using `target_include_directories`/`include_directories`.

1a1a11a committed:
remove cmake files that are not needed

1a1a11a commented:
I think with the current structure, we include the libCacheSim header path `include/libCacheSim` for every target, and we add target_include_directory for others if needed. 
The CMakeLists in admission is not needed any more (and just removed in the new commit), it is merged into a higher-level CMakeLists, which reduces the number of lib and avoids circular dependency.

1a1a11a committed:
remove cmake redundant lib

1a1a11a committed:
fix an issue in rebase
